# noiseOnTree

This repository contains core codes for our paper on “Phylogenetic comparative analysis of single-cell transcriptomes reveals constrained accumulation of gene expression heterogeneity during clonal expansion”. 

1.For lineage tree construction through raw reads of PacBio, please go to PacBio_Analysis_scripts and PacBio_analysis_pipeline files.

   For automatic running, please go to PacBio_analysis_pipeline files. 

   For manual running, please go to PacBio_Analysis_scripts files. 
   Using "python [name of the script] -h" to view the helpfile for each script.

2.Fig2
  
  The final datasets in "datafig2e-f" of "fig2e-f" were generated by "fig2e-f.R" (line1-line410). For finally ploting, plese go to line 412.
  
  fig2g

3.Fig4
  
  The final datasets in "datafig4b-c" of "fig4b-c" were generated by "fig4b-c.R" (line1-line237). For finally ploting, plese go to line 240.
  
  The final datasets in "datafig4d-e" of "fig4d-e" were generated by "fig4d-e.R" (RNA velocity: line1-line254; OU: line257-line393). For finally ploting, plese go to     line 396.

4.Fig5

  The co-fluctuating gene pairs datasets of "fig5" were generated by "fig5.R" (pic.cor: line1-line73; original.cor: line76-line141)

